encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -145,6 +160,109 @@
+  }
+
+  private <T> T decodePreDefinedObject(TypeToken.ClassToken<T> classToken) throws IOException {
+    Class<T> clazz = classToken.getRawType();
+    if (Map.class.isAssignableFrom(clazz)) {
+      @SuppressWarnings("unchecked")
+      T t = (T) decodeMap((TypeToken.ClassToken<Map>) classToken);
+      return t;
+    } else {
+      @SuppressWarnings("unchecked")
+      T t = (T) decodeCollection((TypeToken.ClassToken<Collection>) classToken);
+      return t;
+    }
+  }
+
+  private <K, V, T extends Map<K, V>> T decodeMap(TypeToken.ClassToken<T> classToken)
+      throws IOException {
+    TypeToken<K> keyTypeToken = classToken.getTypeArguments().at(0);
+    TypeToken<V> valTypeToken = classToken.getTypeArguments().at(1);
+    if (!isSingleValue(keyTypeToken))
+      throw new IllegalArgumentException(keyTypeToken + " cannot be used as Map key.");
+
+    T map = newInstance(classToken);
+    reader.beginObject();
+    while (reader.hasNext()) {
+      String keyLiteral = reader.nextName();
+      K key = getActualKey(keyLiteral, keyTypeToken.getRawType());
+      V val = decode(valTypeToken);
+      V replaced = map.put(key, val);
+      if (replaced != null) {
+        throw new IllegalArgumentException("duplicate key: " + key);
+      }
+    }
+    reader.endObject();
+
+    return map;
+  }
+
+  private <E, T extends Collection<E>> T decodeCollection(TypeToken.ClassToken<T> classToken)
+      throws IOException {
+    TypeToken<E> componentTypeToken = classToken.getTypeArguments().at(0);
+    T collection = newInstance(classToken);
+    reader.beginArray();
+    while (reader.hasNext()) {
+      collection.add(decode(componentTypeToken));
+    }
+    reader.endArray();
+    return collection;
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> T newInstance(TypeToken.ClassToken<T> classToken) {
+    Class<T> clazz = classToken.getRawType();
+    if (clazz.equals(HashMap.class) || clazz.equals(Map.class)) {
+      return (T) new HashMap<>();
+    } else if (clazz.equals(LinkedHashMap.class)) {
+      return (T) new LinkedHashMap<>();
+    } else if (clazz.equals(TreeMap.class) || clazz.equals(SortedMap.class)) {
+      return (T) new TreeMap<>();
+    } else if (clazz.equals(ArrayList.class) || clazz.equals(List.class)) {
+      return (T) new ArrayList<>();
+    } else if (clazz.equals(ArrayDeque.class)
+        || clazz.equals(Deque.class)
+        || clazz.equals(Queue.class)) {
+      return (T) new ArrayDeque<>();
+    } else if (clazz.equals(PriorityQueue.class)) {
+      return (T) new PriorityQueue<>();
+    } else if (clazz.equals(HashSet.class) || clazz.equals(Set.class)) {
+      return (T) new HashSet<>();
+    } else if (clazz.equals(LinkedHashSet.class)) {
+      return (T) new LinkedHashSet<>();
+    } else if (clazz.equals(TreeSet.class) || clazz.equals(SortedSet.class)) {
+      return (T) new TreeSet<>();
+    }
+    throw new IllegalArgumentException(classToken + " not supported.");
```

Comment:
Consider throwing `EncodingException` instead as this is the exception we advertise can be thrown.